zetaclient/chains/solana/observer/inbound_parser.go

```diff
@@ -0,0 +1,230 @@
+				Index:            instructionIndex,
+				CoinType:         coin.CoinType_ERC20,
+				Asset:            depositSPL.Asset,
+				IsCrossChainCall: depositSPL.IsCrossChainCall,
+				RevertOptions:    depositSPL.RevertOptions,
+			})
+			p.logger.Info().
+				Str(logs.FieldMethod, "parseInstruction").
+				Str("signature", p.tx.Signatures[0].String()).
+				Str("location", location).
+				Uint32("instructionIndex", instructionIndex).
+				Msg("SPL deposit detected")
+			return nil
+		}
+	} else {
+		p.logger.Warn().
+			Str(logs.FieldMethod, "parseInstruction").
+			Str("signature", p.tx.Signatures[0].String()).
+			Str("location", location).
+			Msg("multiple SPL deposits detected")
+	}
+
+	// try parsing the instruction as a 'call' if not seen yet
+	if !p.seenCall {
+		call, err := solanacontracts.ParseInboundAsCall(p.tx, instruction, p.txResult.Slot)
+		if err != nil {
+			return errors.Wrap(err, "error ParseInboundAsCall")
+		}
+		if call != nil {
+			p.seenCall = true
+			p.events = append(p.events, &clienttypes.InboundEvent{
+				SenderChainID:    p.senderChainID,
+				Sender:           call.Sender,
+				Receiver:         call.Receiver,
+				TxOrigin:         call.Sender,
+				Amount:           call.Amount,
+				Memo:             call.Memo,
+				BlockNumber:      call.Slot,
+				TxHash:           p.tx.Signatures[0].String(),
+				Index:            instructionIndex,
+				CoinType:         coin.CoinType_NoAssetCall,
+				Asset:            call.Asset,
+				IsCrossChainCall: call.IsCrossChainCall,
+				RevertOptions:    call.RevertOptions,
+			})
+			p.logger.Info().
+				Str(logs.FieldMethod, "parseInstruction").
+				Str("signature", p.tx.Signatures[0].String()).
+				Str("location", location).
+				Uint32("instructionIndex", instructionIndex).
+				Msg("call detected")
+			return nil
+		}
+	} else {
+		p.logger.Warn().
+			Str(logs.FieldMethod, "parseInstruction").
+			Str("signature", p.tx.Signatures[0].String()).
+			Str("location", location).
+			Msg("multiple calls detected")
+	}
+
+	return nil
+}
+
+// Parse parses all instructions in the transaction and returns the detected events
+func (p *InboundEventParser) Parse() error {
+	// there should be at least one instruction and one account, otherwise skip
+	if len(p.tx.Message.Instructions) == 0 {
+		return nil
+	}
+
+	// parse top-level instructions
+	for i, instruction := range p.tx.Message.Instructions {
+		if err := p.parseInstruction(instruction, fmt.Sprintf("top-level instruction %d", i), uint32(i)); err != nil {
+			return err
+		}
+	}
+
+	// parse inner instructions
+	if p.txResult.Meta != nil && p.txResult.Meta.InnerInstructions != nil {
+		for _, inner := range p.txResult.Meta.InnerInstructions {
+			for j, instruction := range inner.Instructions {
+				// For inner instructions, we use the outer instruction index as the base
+				// and add the inner instruction index to ensure uniqueness
+				instructionIndex := uint32(inner.Index) + uint32(j)
```

Comment:
Nothing but a question: will this `instructionIndex` be conflicting with above `uint32(i))` in the top-level instruction loop?